File: src/physics_piano/core/voice.py
"""PianoVoice: Unison string group, shared hammer, and individual damper control."""

import math
from typing import Tuple, List
from physics_piano.params.schema import KeyParameters, StringPhysicalParameters
from physics_piano.core.string import StiffStringModal
from physics_piano.core.hammer import HuntCrossleyHammer


class PianoVoice:
    """A single piano key containing unison strings (1 to 3), hammer, and damper."""

    def __init__(self, key_params: KeyParameters, sample_rate: float = 48000.0):
        self.key_params = key_params
        self.sample_rate = float(sample_rate)
        self.dt = 1.0 / self.sample_rate

        self.midi_note = key_params.midi_note
        self.pitch_name = key_params.pitch_name
        self.target_f0 = key_params.target_f0

        # Stereo pan position (0.1=bass/left, 0.9=treble/right)
        self.pan = (self.midi_note - 21) / (108 - 21) * 0.8 + 0.1

        # Instantiate unison strings with micro-detuning
        self.strings: List[StiffStringModal] = []
        for i, s_param in enumerate(key_params.strings):
            cents_detune = key_params.detuning_cents[i] if i < len(key_params.detuning_cents) else 0.0
            # Tension adjustment for micro-detuning: f ~ sqrt(T), so delta_T ~ 2 * delta_f
            freq_ratio = 2.0 ** (cents_detune / 1200.0)
            detuned_tension = s_param.tension * (freq_ratio ** 2)

            detuned_param = StringPhysicalParameters(
                length=s_param.length,
                radius=s_param.radius,
                density=s_param.density,
                youngs_modulus=s_param.youngs_modulus,
                tension=detuned_tension,
                sigma0=s_param.sigma0,
                sigma1=s_param.sigma1,
                strike_ratio=s_param.strike_ratio,
                num_modes=s_param.num_modes,
                polarization_mistuning=s_param.polarization_mistuning
            )
            self.strings.append(StiffStringModal(detuned_param, self.sample_rate))

        # Shared hammer
        self.hammer = HuntCrossleyHammer(key_params.hammer, self.sample_rate)

        self.is_key_down = False
        self.is_sounding = False

    def note_on(self, velocity: float):
        """Depress key and strike unison strings."""
        self.is_key_down = True
        self.is_sounding = True
        # Raise dampers on this voice
        for s in self.strings:
            s.set_damper(False)
        
        # Calculate current average displacement of strings under felt
        u_cur = sum(s.get_strike_displacement_and_velocity()[0] for s in self.strings) / len(self.strings)
        # Strike hammer with current string position context for smooth restrike
        self.hammer.strike(velocity, initial_u_string=u_cur)

    def note_off(self, sustain_pedal: bool = False):
        """Release key. If sustain pedal is not held, lower dampers."""
        self.is_key_down = False
        if not sustain_pedal:
            for s in self.strings:
                s.set_damper(True, depth=1.0)

    def set_sustain_pedal(self, pedal_down: bool):
        """Update damper state according to global sustain pedal."""
        if pedal_down:
            # Pedal up-lifts all dampers
            for s in self.strings:
                s.set_damper(False)
        else:
            # If pedal released and key is not held down, lower dampers
            if not self.is_key_down:
                for s in self.strings:
                    s.set_damper(True, depth=1.0)

    def step(self, f_coupling_T: float = 0.0, f_coupling_P: float = 0.0) -> Tuple[float, float]:
        """Compute one step of hammer interaction and unison string states.
        
        Returns:
            f_bridge_T_sum, f_bridge_P_sum: Forces exerted by this key's unison strings on bridge.
        """
        # 1. Average string displacement and velocity under hammer felt
        u_avg = 0.0
        v_avg = 0.0
        num_str = len(self.strings)
        for s in self.strings:
            u_i, v_i = s.get_strike_displacement_and_velocity()
            u_avg += u_i
            v_avg += v_i
        u_avg /= num_str
        v_avg /= num_str

        # 2. Compute nonlinear hammer force
        f_hammer = self.hammer.compute_force(u_avg, v_avg)
        self.hammer.advance(f_hammer)

        # Split hammer force equally among unison strings
        f_hammer_per_string = f_hammer / num_str

        # 3. Advance each unison string and aggregate bridge boundary forces
        total_bridge_T = 0.0
        total_bridge_P = 0.0
        coupling_per_string_T = f_coupling_T / num_str
        coupling_per_string_P = f_coupling_P / num_str

        for s in self.strings:
            s.step(f_hammer_per_string, coupling_per_string_T, coupling_per_string_P)
            fb_T, fb_P = s.get_bridge_forces()
            total_bridge_T += fb_T
            total_bridge_P += fb_P

        return total_bridge_T, total_bridge_P

File: crates/physics-piano/src/core/voice.rs
//! Piano Voice with Unison String Triplet, Micro-Detuning, and Damper Control in Rust.

use crate::params::KeyParams;
use crate::core::string::StiffStringModal;
use crate::core::hammer::HuntCrossleyHammer;

pub struct PianoVoice {
    pub key_params: KeyParams,
    pub midi_note: u8,
    pub pitch_name: String,
    pub target_f0: f64,
    pub pan: f64,
    pub pan_l: f64,
    pub pan_r: f64,

    pub strings: Vec<StiffStringModal>,
    pub hammer: HuntCrossleyHammer,

    pub is_key_down: bool,
    pub is_sounding: bool,
}

impl PianoVoice {
    pub fn new(key_params: KeyParams, sample_rate: f64) -> Self {
        let midi_note = key_params.midi_note;
        let pitch_name = key_params.pitch_name.clone();
        let target_f0 = key_params.target_f0;
        let pan = ((midi_note as f64 - 21.0) / (108.0 - 21.0) * 0.8 + 0.1).clamp(0.05, 0.95);
        let pan_l = ((1.0 - pan) * std::f64::consts::PI * 0.5).sin();
        let pan_r = (pan * std::f64::consts::PI * 0.5).sin();

        let has_damper = midi_note < 89;
        let mut strings = Vec::with_capacity(key_params.num_unisons);
        for (i, s_param) in key_params.strings.iter().enumerate() {
            let cents_detune = if i < key_params.detuning_cents.len() {
                key_params.detuning_cents[i]
            } else {
                0.0
            };
            let freq_ratio = 2.0f64.powf(cents_detune / 1200.0);
            let detuned_tension = s_param.tension * freq_ratio.powi(2);

            let mut detuned_param = s_param.clone();
            detuned_param.tension = detuned_tension;

            let mut s = StiffStringModal::new(detuned_param, sample_rate);
            s.has_damper = has_damper;
            if !has_damper {
                s.set_damper(false, 0.0);
                s.current_damper_depth = 0.0;
            }
            strings.push(s);
        }

        let hammer = HuntCrossleyHammer::new(key_params.hammer.clone(), sample_rate);

        Self {
            key_params,
            midi_note,
            pitch_name,
            target_f0,
            pan,
            pan_l,
            pan_r,
            strings,
            hammer,
            is_key_down: false,
            is_sounding: false,
        }
    }

    pub fn reset(&mut self) {
        self.is_sounding = false;
        self.is_key_down = false;
        self.hammer.reset();
        for s in &mut self.strings {
            s.reset();
        }
    }

    pub fn set_tuning_offset(&mut self, cents: f64) {
        for s in &mut self.strings {
            s.set_tuning_offset(cents);
        }
    }

    pub fn set_una_corda(&mut self, enabled: bool) {
        self.hammer.set_una_corda(enabled);
    }

    pub fn set_hammer_hardness(&mut self, scale: f64) {
        let s = scale.clamp(0.4, 3.0);
        self.hammer.k_h = self.key_params.hammer.stiffness * s.powf(2.0);
        self.hammer.p = (self.key_params.hammer.exponent * s.sqrt()).clamp(1.5, 3.5);
    }

    pub fn set_unison_detuning(&mut self, detune_scale: f64) {
        for (i, s) in self.strings.iter_mut().enumerate() {
            let base_cents = if i < self.key_params.detuning_cents.len() {
                self.key_params.detuning_cents[i]
            } else {
                0.0
            };
            s.set_tuning_offset(base_cents * detune_scale);
        }
    }

    pub fn set_inharmonicity_scale(&mut self, inharm_scale: f64) {
        for s in &mut self.strings {
            s.set_inharmonicity_scale(inharm_scale);
        }
    }

    pub fn set_damper_depth(&mut self, depth: f64) {
        let active = depth > 0.0;
        for s in &mut self.strings {
            s.set_damper(active, depth);
        }
    }

    pub fn get_energy(&self) -> f64 {
        let e_strings: f64 = self.strings.iter().map(|s| s.get_energy()).sum();
        let e_hammer = if self.hammer.is_active {
            0.5 * self.hammer.m_h * self.hammer.v_h.powi(2)
        } else {
            0.0
        };
        e_strings + e_hammer
    }

    pub fn note_on(&mut self, velocity: f64) {
        self.is_key_down = true;
        self.is_sounding = true;
        for s in &mut self.strings {
            s.set_damper(false, 0.0);
            s.current_damper_depth = 0.0;
        }

        let u_avg: f64 = self.strings.iter().map(|s| s.get_strike_displacement_and_velocity().0).sum::<f64>()
            / self.strings.len() as f64;
        self.hammer.strike(velocity, u_avg);
    }

    pub fn update_damper_state(&mut self, sustain_pedal: bool, pedal_depth: f64) {
        // 1. If key is physically held down by pianist's finger,
        // the whippen/damper lever keeps the damper 100% lifted off the string!
        if self.is_key_down {
            for s in &mut self.strings {
                s.set_damper(false, 0.0);
            }
            return;
        }

        // 2. Key is released: damper position is governed by sustain pedal
        // Acoustic Grand Piano damper lift curve:
        // - depth in [0.0, 0.20]: Lost motion (pedal clearance), dampers remain fully seated (effective damping = 1.0)
        // - depth in [0.20, 0.70]: Half-pedal zone, dampers gradually lift off string
        // - depth in [0.70, 1.00]: Dampers completely clear off string (effective damping = 0.0, fully sustained)
        if sustain_pedal && pedal_depth > 0.20 {
            if pedal_depth >= 0.70 {
                // Fully lifted: sustain
                for s in &mut self.strings {
                    s.set_damper(false, 0.0);
                }
            } else {
                // Half-pedal zone: progressive damping
                let norm = (pedal_depth - 0.20) / 0.50; // 0.0 to 1.0
                let effective_damping = (1.0 - norm).powi(2);
                if effective_damping < 0.02 {
                    for s in &mut self.strings {
                        s.set_damper(false, 0.0);
                    }
                } else {
                    for s in &mut self.strings {
                        s.set_damper(true, effective_damping);
                    }
                }
            }
        } else {
            // Pedal released: full damping
            for s in &mut self.strings {
                s.set_damper(true, 1.0);
            }
        }
    }

    pub fn note_off(&mut self, sustain_pedal: bool, pedal_depth: f64) {
        self.is_key_down = false;
        self.update_damper_state(sustain_pedal, pedal_depth);
    }

    pub fn set_sustain_pedal(&mut self, pedal_down: bool, depth: f64) {
        self.update_damper_state(pedal_down, depth);
    }

    #[inline]
    pub fn step(&mut self, f_coupling_t: f64, f_coupling_p: f64) -> (f64, f64, f64) {
        let num_str = self.strings.len() as f64;
        let is_una_corda = self.hammer.una_corda;
        // Physical shift: for multi-string unisons (triplets/bichords), shift drops one string
        let num_struck = if is_una_corda && self.strings.len() >= 2 {
            self.strings.len() - 1
        } else {
            self.strings.len()
        };

        let f_hammer = if self.hammer.is_active {
            let mut u_avg = 0.0;
            let mut v_avg = 0.0;
            for s in &self.strings[..num_struck] {
                let (u, v) = s.get_strike_displacement_and_velocity();
                u_avg += u;
                v_avg += v;
            }
            u_avg /= num_struck as f64;
            v_avg /= num_struck as f64;

            let f_h = self.hammer.compute_force(u_avg, v_avg);
            self.hammer.advance(f_h);
            f_h
        } else {
            0.0
        };

        let f_hammer_per_struck = f_hammer / num_struck as f64;
        let coupling_per_string_t = f_coupling_t / num_str;
        let coupling_per_string_p = f_coupling_p / num_str;

        let mut total_bridge_t = 0.0;
        let mut total_bridge_p = 0.0;
        let mut total_bridge_l = 0.0;

        for (i, s) in self.strings.iter_mut().enumerate() {
            let f_h_i = if i < num_struck { f_hammer_per_struck } else { 0.0 };
            let (fb_t, fb_p, fb_l) = s.step(f_h_i, coupling_per_string_t, coupling_per_string_p);
            total_bridge_t += fb_t;
            total_bridge_p += fb_p;
            total_bridge_l += fb_l;
        }

        (total_bridge_t, total_bridge_p, total_bridge_l)
    }
}
